README.md
[![Review Assignment Due Date](https://classroom.github.com/assets/deadline-readme-button-24ddc0f5d75046c5622901739e7c5dd533143b0c8e959d652212380cedb1ea36.svg)](https://classroom.github.com/a/q2MA-8aZ)


CHANGELOG.md
# Changelog for `haskell-boar`

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to the
[Haskell Package Versioning Policy](https://pvp.haskell.org/).

## Unreleased

## 0.1.0.0 - YYYY-MM-DD


src/Game.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE ImportQualifiedPost #-}
{-# LANGUAGE OverloadedStrings #-}
{-# OPTIONS_GHC -Wno-partial-fields #-}

module Game
  ( State (..),
    World (..),
    Player (..),
    Enemy (..),
    Entity (..),
    Stats (..),
    Action (..),
    Map (..),
    Tile (..),
    Mode (..),
    Damageable (..),
    EnemyState (..),
    Scene (..),
    isValidAction,
    newState,
    isMapBounded,
    (!!!),
    createMaps,
    getTileFromName,
    isTileWalkable,
    findTilesInDistance,
    findWalkableTilesInDistance,
    getMapEnemies,
    isEntityInTile,
    sumEnemiesAttack,
    pointDiff,
    manhattanDist,
    restartGame,
  )
where

import Data.Aeson hiding (Key)
import Data.Map qualified as M
import Data.Set qualified as S
import GHC.Generics
import Graphics.Gloss.Interface.IO.Interact (Key (..), Picture, Point)

data Scene = Menu | Game | GameOver | Win deriving (Show, Eq)

data State a = State
  { sData :: a,
    sKeys :: S.Set Key,
    updateTimer :: Float,
    playerAction :: Action,
    lastMousePosition :: Point,
    showAttackAnimation :: Bool,
    currentScene :: Scene
  }

data Action = NoAction | Move Point | Attack Point deriving (Show, Eq)

isValidAction :: Action -> Bool
isValidAction NoAction = False
isValidAction _ = True

data Stats = Stats
  { maxLife :: Int,
    life :: Int,
    attack :: Int
  }
  deriving (Show)

data Entity = Entity
  { ePos :: Point,
    eStats :: Stats,
    eTexture :: Picture
  }
  deriving (Show)

pointDiff :: Point -> Point -> Point
pointDiff (originX, originY) (targetX, targetY) = (targetX - originX, targetY - originY)

manhattanDist :: Point -> Int
manhattanDist (x, y) = abs (round x) + abs (round y)

newEntity :: Point -> Int -> Int -> Picture -> Entity
newEntity startPos startLife attackValue = Entity startPos (Stats startLife startLife attackValue)

data Player = Player {pEnt :: Entity, pMaxMoveDistance :: Int, pMaxAttackDistance :: Int}
  deriving (Show)

data EnemyState = EIdle | EFollow | EAttack | ERangedAttack deriving (Show, Eq)

data Enemy
  = Melee {eEnt :: Entity, eState :: EnemyState}
  | Ranged {eEnt :: Entity, eState :: EnemyState}
  deriving (Show, Generic)

class Damageable a where
  damage :: a -> Int -> Maybe a

instance Damageable Entity where
  damage entity value
    | newLife <= 0 = Nothing
    | otherwise = Just $ entity {eStats = stats'}
    where
      stats = eStats entity
      newLife = life stats - value
      stats' = stats {life = life stats - value}

sumEnemiesAttack :: [Enemy] -> Int
sumEnemiesAttack mapEnemies =
  sum
    [ attack $ eStats $ eEnt enemy
      | enemy <- mapEnemies,
        eState enemy == EAttack || eState enemy == ERangedAttack
    ]

data MapEnemyDefinition = MapEnemyDefinition
  { enemyType :: String,
    enemyLife :: Int,
    enemyAttack :: Int,
    enemyPos :: Point
  }
  deriving (Show, Generic)

instance ToJSON MapEnemyDefinition where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON MapEnemyDefinition where
  parseJSON = withObject "MapEnemyDefinition" $ \o -> do
    type_ <- o .: "type"
    enemyLifeDef <- o .: "life"
    enemyAttackDef <- o .: "attack"
    pos <- o .: "pos"
    return MapEnemyDefinition {enemyType = type_, enemyLife = enemyLifeDef, enemyAttack = enemyAttackDef, enemyPos = pos}

createEnemyFromDefinition :: State World -> MapEnemyDefinition -> Enemy
createEnemyFromDefinition state def = case enemyType def of
  "melee" -> Melee (newEntity (enemyPos def) (enemyLife def) (enemyAttack def) (head textures)) EIdle
  "ranged" -> Ranged (newEntity (enemyPos def) (enemyLife def) (enemyAttack def) (textures !! 1)) EIdle
  _ -> undefined
  where
    textures = wEnemyPictures $ sData state

data Map a = Map
  { mTiles :: [[a]],
    mEnemies :: [MapEnemyDefinition],
    mPlayerPos :: Point
  }
  deriving (Show, Generic)

instance (ToJSON a) => ToJSON (Map a) where
  toEncoding = genericToEncoding defaultOptions

instance (FromJSON a) => FromJSON (Map a) where
  parseJSON = withObject "Map" $ \o -> do
    jsonTiles <- o .: "tiles"
    mapEnemies <- o .: "enemies"
    playerPos <- o .: "playerPos"

    return Map {mTiles = jsonTiles, mEnemies = mapEnemies, mPlayerPos = playerPos}

instance Functor Map where
  fmap f Map {mTiles = tiles, mEnemies = mapEnemies, mPlayerPos = playerPos} =
    Map {mTiles = map (map f) tiles, mEnemies = mapEnemies, mPlayerPos = playerPos}

instance Applicative Map where
  pure a = Map [[a]] [] (0, 0)
  Map [[f]] _ _ <*> Map values eCount pPos = Map (map (map f) values) eCount pPos
  Map fs _ _ <*> Map values eCount pPos = Map (zipWith (zipWith ($)) fs values) eCount pPos

data Tile
  = Tile
      { tName :: String,
        tTexture :: String,
        tWalkable :: Bool
      }
  | EmptyTile
  deriving (Show, Generic)

isTileWalkable :: Tile -> Bool
isTileWalkable (Tile _ _ walkable) = walkable
isTileWalkable EmptyTile = False

instance ToJSON Tile where
  toEncoding (Tile name texture walkable) = pairs ("name" .= name <> "texture" .= texture <> "walkable" .= walkable)
  toEncoding EmptyTile = error "EmptyTile is not serializable"

instance FromJSON Tile where
  parseJSON = withObject "Tile" $ \o -> do
    name <- o .: "name"
    texture <- o .: "texture"
    walkable <- o .: "isWalkable"
    return Tile {tName = name, tTexture = texture, tWalkable = walkable}

data Mode = NoMode | MoveMode [Point] | AttackMode [Point] | EnemyMode deriving (Show, Eq)

data World = World
  { wPlayer :: Player,
    wTiles :: [Tile],
    wPictureTileMap :: M.Map String Picture,
    wMaps :: [Map Tile],
    wCurrentMap :: Int,
    wEnemies :: [Enemy],
    wMode :: Mode,
    wSword :: Picture,
    wFireball :: Picture,
    wEnemyPictures :: [Picture]
  }
  deriving (Show)

(!!!) :: Map a -> Point -> a
Map {mTiles = tiles} !!! (x, y) = (tiles !! floor y) !! floor x

isMapBounded :: Map a -> Point -> Bool
isMapBounded map' (x, y)
  | x < 0 || x >= fromIntegral width = False
  | y < 0 || y >= fromIntegral height = False
  | otherwise = True
  where
    height = length $ mTiles map'
    width = length $ head $ mTiles map'

isEnemyInTile :: World -> Point -> Bool
isEnemyInTile world tile = tile `elem` map (ePos . eEnt) (wEnemies world)

isEntityInTile :: World -> Point -> Bool
isEntityInTile world tile = isEnemyInTile world tile || tile == ePos (pEnt $ wPlayer world)

findTilesInDistance :: Map Tile -> Point -> Float -> [Point]
findTilesInDistance map' (px, py) distance =
  filter
    (isMapBounded map')
    [ (x, y)
      | x <- [px - distance .. px + distance],
        y <- [py - distance .. py + distance],
        abs (x - px) + abs (y - py) <= distance
    ]

findWalkableTilesInDistance :: Map Tile -> Point -> Float -> [Point]
findWalkableTilesInDistance map' (px, py) distance =
  filter
    (\tilePos -> isTileWalkable (map' !!! tilePos))
    (findTilesInDistance map' (px, py) distance)

getMapEnemies :: State World -> Map a -> [Enemy]
getMapEnemies state map' = map (createEnemyFromDefinition state) (mEnemies map')

newState :: (Picture, Picture, Picture) -> [Picture] -> [Map Tile] -> M.Map String Picture -> [Tile] -> State World
newState (playerPicture, sword, fireball) enemyPictures maps pictureMap tiles = state {sData = world {wEnemies = mapEnemies}}
  where
    state =
      State
        { sData =
            World
              { wPlayer =
                  Player (newEntity playerPos 10 2 playerPicture) 2 1,
                wEnemies = [],
                wTiles = tiles,
                wPictureTileMap = pictureMap,
                wCurrentMap = 0,
                wMaps = maps,
                wMode = NoMode,
                wSword = sword,
                wFireball = fireball,
                wEnemyPictures = enemyPictures
              },
          updateTimer = 0.0,
          playerAction = NoAction,
          lastMousePosition = (0, 0),
          showAttackAnimation = False,
          sKeys = S.empty,
          currentScene = Menu
        }
    world = sData state
    playerPos = mPlayerPos $ head maps
    mapEnemies = getMapEnemies state (head maps)

getTileFromName :: String -> [Tile] -> Tile
getTileFromName name tiles = head $ filter (\tile -> tName tile == name) tiles

restartGame :: State World -> State World
restartGame state = state {sData = world {wPlayer = player', wCurrentMap = currentMap', wEnemies = enemies'}, currentScene = Game}
  where
    world = sData state
    player' = Player (newEntity playerPos 10 2 (eTexture $ pEnt $ wPlayer world)) 2 1
    currentMap' = 0
    enemies' = getMapEnemies state (head $ wMaps world)
    playerPos = mPlayerPos $ head $ wMaps world

createMaps :: [Map Char] -> [Tile] -> [Map Tile]
createMaps charMaps tiles = map createMap charMaps
  where
    createTile :: Char -> Tile
    createTile char = case char of
      'D' -> getTileFromName "Dirt" tiles
      'W' -> getTileFromName "Water" tiles
      ' ' -> EmptyTile
      _ -> undefined

    createMap :: Map Char -> Map Tile
    createMap = fmap createTile


src/Main.hs
{-# LANGUAGE ImportQualifiedPost #-}

module Main (main) where

import Config
import Data.Aeson (decode)
import Data.ByteString.Lazy qualified as BSL
import Data.Map qualified as M
import Events (handleGameEvents, handleGameOverEvents, handleMenuEvents, handleWinEvents)
import Game
import Graphics.Gloss
import Graphics.Gloss.Interface.IO.Game
import Renderer
  ( renderAttackAnimation,
    renderEnemies,
    renderGameOver,
    renderHUD,
    renderMainMenu,
    renderMap,
    renderPlayer,
    renderPossibleMoves,
    renderWin,
  )
import System.Directory (getDirectoryContents)
import System.FilePath ((</>))
import Update

render :: State World -> Picture
render state
  | scene == Menu = renderMainMenu state
  | scene == Game = pictures [scale scalingFactor scalingFactor $ pictures renderAll, renderHUD state]
  | scene == GameOver = renderGameOver state
  | scene == Win = renderWin state
  | otherwise = blank
  where
    scene = currentScene state
    renderAll =
      map
        (\f -> f state)
        [ renderMap,
          renderPossibleMoves,
          renderPlayer,
          renderEnemies,
          renderAttackAnimation
        ]

handleEvents :: Event -> State World -> State World
handleEvents event state
  | scene == Menu = handleMenuEvents event state
  | scene == Game = handleGameEvents event state
  | scene == GameOver = handleGameOverEvents event state
  | scene == Win = handleWinEvents event state
  | otherwise = state
  where
    scene = currentScene state

update :: Float -> State World -> State World
update dt state
  | scene == Game = updateGame dt state
  | otherwise = state
  where
    scene = currentScene state

isFile :: FilePath -> Bool
isFile path = path /= "." && path /= ".."

main :: IO ()
main =
  do
    tilesContent <- getDirectoryContents tilesFolder
    let tiles = map (tilesFolder </>) $ filter isFile tilesContent

    gameTiles <- mapM loadTile tiles
    tilePictures <- mapM (loadBMP . tTexture) gameTiles

    let zippedTiles = zip (map tTexture gameTiles) tilePictures
    let tileMap = M.fromList zippedTiles

    mapsContent <- getDirectoryContents mapsFolder
    let maps = map (mapsFolder </>) $ filter isFile mapsContent

    charMaps <- mapM loadMap maps
    let gameMaps = createMaps charMaps gameTiles

    playerPicture <- loadBMP (charactersFolder </> "knight.bmp")

    enemyPicture1 <- loadBMP (charactersFolder </> "ogre.bmp")
    enemyPicture2 <- loadBMP (charactersFolder </> "mage.bmp")

    swordPicture <- loadBMP (itemsFolder </> "sword.bmp")
    fireballPicture <- loadBMP (objectsFolder </> "fireball.bmp")

    let window = InWindow "My Window" (screenWidth, screenHeight) (100, 100)
    let initialState = newState (playerPicture, swordPicture, fireballPicture) [enemyPicture1, enemyPicture2] gameMaps tileMap gameTiles

    play window black fps initialState render handleEvents update
  where
    loadTile tilePath = do
      tileJson <- BSL.readFile tilePath
      let tile = decode tileJson :: Maybe Tile

      case tile of
        Just t -> return t
        Nothing -> error $ "Could not load tile " ++ tilePath

    loadMap mapPath = do
      mapContent <- BSL.readFile mapPath
      let map_ = decode mapContent :: Maybe (Map Char)

      case map_ of
        Just m -> return m
        Nothing -> error $ "Could not load map " ++ mapPath


src/Events.hs
module Events (handleMenuEvents, handleGameEvents, handleGameOverEvents, handleWinEvents) where

import Config (updateInterval)
import Game (Action (Attack, Move), Entity (ePos), Mode (AttackMode, MoveMode, NoMode), Player (pEnt, pMaxAttackDistance, pMaxMoveDistance), Scene (..), State (..), World (..), findWalkableTilesInDistance, isEntityInTile, restartGame)
import Graphics.Gloss.Interface.IO.Game
import Renderer (screenPositionToWorldPosition)

handleMenuEvents :: Event -> State World -> State World
handleMenuEvents (EventKey (SpecialKey KeySpace) Down _ _) state =
  state {currentScene = Game}
handleMenuEvents _ state = state

handleGameOverEvents :: Event -> State World -> State World
handleGameOverEvents (EventKey (SpecialKey KeySpace) Down _ _) state = restartGame state
handleGameOverEvents _ state = state

handleWinEvents :: Event -> State World -> State World
handleWinEvents (EventKey (SpecialKey KeySpace) Down _ _) state = restartGame state
handleWinEvents _ state = state

handleGameEvents :: Event -> State World -> State World
handleGameEvents event state
  | updateTimer state < updateInterval = state
  | otherwise = case event of
      EventKey (Char 'm') Down _ _ ->
        let mode = wMode world
            playerPos = ePos $ pEnt $ wPlayer world
            maxDistance = fromIntegral $ pMaxMoveDistance $ wPlayer world
            walkableTilesInMoveRange =
              filter
                (not . isEntityInTile world)
                (findWalkableTilesInDistance currentMap playerPos maxDistance)
         in case mode of
              MoveMode _ -> state {sData = world {wMode = NoMode}}
              _ -> state {sData = world {wMode = MoveMode walkableTilesInMoveRange}}
      EventKey (Char 'a') Down _ _ ->
        let mode = wMode world
            playerPos = ePos $ pEnt $ wPlayer world
            maxDistance = fromIntegral $ pMaxAttackDistance $ wPlayer world
            walkableTilesInAttackRange = findWalkableTilesInDistance currentMap playerPos maxDistance
         in case mode of
              AttackMode _ -> state {sData = world {wMode = NoMode}}
              _ -> state {sData = world {wMode = AttackMode walkableTilesInAttackRange}}
      EventKey (MouseButton LeftButton) Down _ (mouseX, mouseY) ->
        let (x, y) = screenPositionToWorldPosition (mouseX, mouseY)
            (px, py) = ePos $ pEnt $ wPlayer world
         in case wMode world of
              MoveMode walkableTiles ->
                if (x, y) `elem` walkableTiles
                  then state {lastMousePosition = (x, y), playerAction = Move (x - px, y - py)}
                  else state {lastMousePosition = (x, y)}
              AttackMode possibleAttackTiles ->
                if (x, y) `elem` possibleAttackTiles
                  then state {lastMousePosition = (x, y), playerAction = Attack (x, y), showAttackAnimation = True}
                  else state {lastMousePosition = (x, y)}
              _ -> state
      _ -> state
  where
    world = sData state
    currentMap = (!! wCurrentMap world) $ wMaps world


src/Update.hs
module Update (updateWorld, refillEntityLife, updateGame) where

import Config (updateInterval)
import Data.Bifunctor (Bifunctor (bimap))
import Data.Maybe (fromMaybe)
import Game
  ( Action (..),
    Damageable (damage),
    Enemy (..),
    EnemyState (EAttack, EFollow, ERangedAttack),
    Entity (..),
    Map (mPlayerPos),
    Mode (EnemyMode, NoMode),
    Player (..),
    Scene (GameOver, Win),
    State (..),
    Stats (..),
    World (..),
    findWalkableTilesInDistance,
    getMapEnemies,
    isEntityInTile,
    isValidAction,
    manhattanDist,
    pointDiff,
    sumEnemiesAttack,
  )
import Graphics.Gloss (Point)

moveEntity :: Point -> Entity -> Entity
moveEntity point entity = entity {ePos = finalPos}
  where
    (px, py) = ePos entity
    finalPos = bimap (px +) (py +) point

refillEntityLife :: Entity -> Entity
refillEntityLife entity = entity {eStats = entityStats {life = refilledLife}}
  where
    entityStats = eStats entity
    refilledLife = maxLife entityStats

updatePlayer :: [Enemy] -> Float -> State World -> Player
updatePlayer enemies _ state = case pAction of
  Move dir -> player {pEnt = if isValidMove dir then movedPlayer dir else playerEntity}
  _ -> player {pEnt = playerEntity}
  where
    pAction = playerAction state
    player = wPlayer $ sData state
    enemiesDamage = sumEnemiesAttack enemies
    movedPlayer dir = moveEntity dir playerEntity
    movedPos dir = ePos $ movedPlayer dir
    isValidMove dir = movedPos dir `notElem` map (ePos . eEnt) enemies

    playerEntity =
      fromMaybe
        (pEnt player) {eStats = (eStats $ pEnt player) {life = 0}}
        (damage (pEnt player) enemiesDamage)

updateEnemies :: Float -> State World -> [Enemy]
updateEnemies _ state = map updateEnemy enemies
  where
    enemies = wEnemies $ sData state

    updateEnemy :: Enemy -> Enemy
    updateEnemy enemy = case enemy of
      Melee _ _ ->
        if distanceToPlayer <= 1
          then enemy {eState = EAttack}
          else enemy {eEnt = moveEntity moveDir enemyEntity, eState = EFollow}
      Ranged _ _ ->
        if distanceToPlayer <= 8
          then enemy {eState = ERangedAttack}
          else enemy {eEnt = moveEntity moveDir enemyEntity}
      where
        enemyEntity = eEnt enemy
        enemyPos = ePos enemyEntity
        playerPos = ePos $ pEnt $ wPlayer $ sData state
        distanceToPlayer = manhattanDist $ pointDiff enemyPos playerPos

        currentMap = (!! wCurrentMap (sData state)) $ wMaps (sData state)
        possibleTilesToWalk = filter (not . isEntityInTile (sData state)) (findWalkableTilesInDistance currentMap enemyPos 1)
        distances = map (\tPos -> (manhattanDist $ pointDiff tPos playerPos, tPos)) possibleTilesToWalk
        moveDir = pointDiff enemyPos closestTile
        closestTile
          | null distances = enemyPos
          | otherwise = snd $ minimum distances

updateEnemyAtPosition :: Point -> [Enemy] -> [Enemy]
updateEnemyAtPosition _ [] = []
updateEnemyAtPosition pos (e : es)
  | ePos (eEnt e) == pos = case entity of
      Just entity' -> e {eEnt = entity'} : es
      Nothing -> es
  | otherwise = e : updateEnemyAtPosition pos es
  where
    entity = damage (eEnt e) 1

updateWorld :: Float -> State World -> World
updateWorld dt state = sData state'
  where
    mode =
      if playerAction state == NoAction
        then NoMode
        else EnemyMode

    player' = updatePlayer enemies dt state

    enemies = updateEnemies dt state
    enemies' = case playerAction state of
      Attack attackPos -> updateEnemyAtPosition attackPos enemies
      _ -> enemies

    state'
      | isValidAction (playerAction state) =
          state
            { sData =
                (sData state)
                  { wPlayer = player',
                    wEnemies = enemies',
                    wMode = mode
                  },
              playerAction = NoAction,
              updateTimer = 0
            }
      | otherwise = state

updateGame :: Float -> State World -> State World
updateGame dt state
  | life (eStats playerEnt) <= 0 = state {currentScene = GameOver}
  | wCurrentMap world >= length (wMaps world) = state {currentScene = Win}
  | isValidAction action = state {sData = updateWorld dt state, updateTimer = 0, playerAction = NoAction}
  | showAttackAnimation state && updateTimer state >= updateInterval = state {showAttackAnimation = False, updateTimer = 0}
  | null $ wEnemies world =
      state
        { sData =
            world
              { wEnemies = newEnemies,
                wCurrentMap = wCurrentMap world + 1,
                wPlayer = (wPlayer world) {pEnt = (refillEntityLife playerEnt) {ePos = playerPos}}
              }
        }
  | otherwise = state {updateTimer = curUpdateTimer, sData = world}
  where
    world = sData state
    playerEnt = pEnt $ wPlayer $ sData state
    action = playerAction state
    curUpdateTimer = updateTimer state + dt
    mapIndex = (wCurrentMap world + 1) `mod` length (wMaps world)
    map' = (!! mapIndex) $ wMaps world
    newEnemies = getMapEnemies state map'
    playerPos = mPlayerPos map'


src/Renderer.hs
{-# LANGUAGE ImportQualifiedPost #-}
{-# HLINT ignore "Use tuple-section" #-}
{-# LANGUAGE TupleSections #-}
{-# OPTIONS_GHC -Wno-type-defaults #-}
{-# OPTIONS_GHC -Wno-unrecognised-pragmas #-}

module Renderer
  ( renderMap,
    renderPlayer,
    renderEnemies,
    screenPositionToWorldPosition,
    renderHUD,
    renderPossibleMoves,
    renderGameModeText,
    renderAttackAnimation,
    renderMainMenu,
    renderGameOver,
    renderWin,
  )
where

import Config (halfTileSize, scalingFactor, tileSize)
import Data.Map qualified as M
import Game
import Graphics.Gloss

screenPositionToWorldPosition :: Point -> Point
screenPositionToWorldPosition (mouseX, mouseY) = (fromIntegral x, fromIntegral y)
  where
    rmx = (fromIntegral . floor) mouseX
    rmy = (fromIntegral . floor) mouseY
    x = floor $ (1 / (scalingFactor * tileSize)) * (rmx / 2 + rmy)
    y = floor $ (1 / (scalingFactor * tileSize)) * (rmy - rmx / 2)

worldPositionToScreenPosition :: Point -> Point
worldPositionToScreenPosition (x, y) = (x', y')
  where
    x' = (x - y) * tileSize
    y' = (x + y) * halfTileSize + tileSize

renderHUD :: State World -> Picture
renderHUD state = pictures $ map (scale 0.2 0.2) [renderActionsHelperText, renderGameModeText state, renderLifeBar player]
  where
    world = sData state
    player = wPlayer world

renderLifeBar :: Player -> Picture
renderLifeBar player = pictures [background, foreground]
  where
    background = color white $ translate (-2150) 1900 $ rectangleSolid 510 110
    foreground = color red $ translate (-2150) 1900 $ rectangleSolid (500 * playerLifeRatio) 100
    playerStats = eStats $ pEnt player
    playerLifeRatio = fromIntegral (life playerStats) / fromIntegral (maxLife playerStats)

renderGameModeText :: State World -> Picture
renderGameModeText state = translate (-2000) (-600) $ color white $ text $ case mode of
  MoveMode _ -> "Move mode"
  AttackMode _ -> "Attack mode"
  _ -> "Waiting for mode selection..."
  where
    world = sData state
    mode = wMode world

renderActionsHelperText :: Picture
renderActionsHelperText =
  pictures
    [ translate (-2000) (-800) $ color white $ text "Press m to enter move mode",
      translate (-2000) (-1000) $ color white $ text "Press a to enter attack mode"
    ]

renderTile :: State World -> (Float, Float, Tile) -> Picture
renderTile state (x, y, Tile _ texPath _) = translate x' y' tex
  where
    world = sData state
    worldTiles = wPictureTileMap world
    x' = (x - y) * tileSize
    y' = (x + y) * halfTileSize
    tex = worldTiles M.! texPath
renderTile _ (_, _, EmptyTile) = pictures []

renderMap :: State World -> Picture
renderMap state
  | wCurrentMap world >= length (wMaps world) = blank
  | otherwise = pictures $ renderTile state <$> tiles
  where
    world = sData state

    zipTiles :: [[Tile]] -> [(Float, Float, Tile)]
    zipTiles = concat . zipWith (\y -> zipWith (\x -> (x,y,)) [0 ..]) [0 ..]

    tiles = reverse $ zipTiles $ mTiles $ (!! wCurrentMap world) $ wMaps world

renderPossibleMoves :: State World -> Picture
renderPossibleMoves state = case mode of
  MoveMode walkableTilesInMoveRange -> renderMove indicatorTile walkableTilesInMoveRange
  AttackMode attackableTilesInAttackRange -> renderMove attackIndicatorTile attackableTilesInAttackRange
  _ -> pictures []
  where
    world = sData state
    mode = wMode world
    indicatorTile = getTileFromName "Indicator" $ wTiles world
    attackIndicatorTile = getTileFromName "AttackIndicator" $ wTiles world
    renderMove tile tileList = pictures $ map (renderTile state . (\(x, y) -> (x, y, tile))) tileList

renderAttackAnimation :: State World -> Picture
renderAttackAnimation state
  | showAttackAnimation state = translate x y $ rotate (timer * 50) sword
  | otherwise = pictures []
  where
    (x, y) = worldPositionToScreenPosition $ lastMousePosition state
    timer = updateTimer state
    world = sData state
    sword = wSword world

renderPlayer :: State World -> Picture
renderPlayer state = renderEntity playerEntity
  where
    world = sData state
    playerEntity = pEnt $ wPlayer world

renderEnemies :: State World -> Picture
renderEnemies state = pictures [attackIndicator, pictures [renderEntity $ eEnt enemy | enemy <- mapEnemies]]
  where
    world = sData state
    mapEnemies = wEnemies world

    rangedEnemies = filter isRanged mapEnemies
    rangedEnemiesInAttackDistance =
      filter
        (\enemy -> manhattanDist (pointDiff (ePos (eEnt enemy)) (ePos $ pEnt $ wPlayer world)) <= 8)
        rangedEnemies

    attackIndicator =
      pictures $
        map
          (\enemy -> let (x, y) = worldPositionToScreenPosition $ ePos $ eEnt enemy in translate x (y + 20) $ wFireball world)
          rangedEnemiesInAttackDistance

    isRanged (Melee _ _) = False
    isRanged _ = True

renderEntity :: Entity -> Picture
renderEntity entity = translate x y texture
  where
    (px, py) = ePos entity
    x = (px - py) * tileSize
    y = (px + py) * halfTileSize + tileSize
    texture = eTexture entity

renderMainMenu :: State World -> Picture
renderMainMenu _ = pictures [title, message]
  where
    title = scale 0.5 0.5 $ translate (-300) 0 $ color white $ text "Tiny Hero"
    message = scale 0.15 0.15 $ translate (-600) (-400) $ color white $ text "Press space to start"

renderGameOver :: State World -> Picture
renderGameOver _ = pictures [title, message]
  where
    title = scale 0.5 0.5 $ translate (-300) 0 $ color white $ text "Game Over"
    message = scale 0.15 0.15 $ translate (-600) (-400) $ color white $ text "Press space to restart"

renderWin :: State World -> Picture
renderWin _ = pictures [title, message]
  where
    title = scale 0.5 0.5 $ translate (-300) 0 $ color white $ text "You Won!"
    message = scale 0.15 0.15 $ translate (-600) (-400) $ color white $ text "Press space to restart"


src/GameObjects.hs
module GameObjects
  ( Player (..),
    Enemy (..),
    Action (..),
    isValidAction,
    newEntity,
    updatePlayer,
  )
where

import Game (Action (..))
import Graphics.Gloss.Interface.IO.Interact (Point)

data Action = NoAction | Move Point | Attack Point deriving (Show)

isValidAction :: Action -> Bool
isValidAction NoAction = False
isValidAction _ = True

data Stats = Stats
  { maxLife :: Integer,
    life :: Integer,
    attack :: Integer
  }
  deriving (Show)

data Entity = Entity
  { ePos :: Point,
    eId :: Integer, -- TODO: Implement unique IDs
    eStats :: Stats
  }
  deriving (Show)

newEntity :: Point -> Integer -> Integer -> Entity
newEntity startPos startLife attack = Entity startPos 0 (Stats startLife startLife attack)

moveEntity :: Entity -> Point -> Entity
moveEntity entity (dx, dy) = entity {ePos = (x + dx, y + dy)}
  where
    pos = ePos entity
    x = fst pos
    y = snd pos

data Player = Player {pEnt :: Entity}
  deriving (Show)

updatePlayer :: Action -> Player -> Player
updatePlayer (Move dir) (Player ent) = Player $ moveEntity ent dir
updatePlayer _ player = player

data Enemy = Melee {eEnt :: Entity} | Ranged {eEnt :: Entity}


src/Config.hs
{-# LANGUAGE OverloadedStrings #-}

module Config
  ( fps,
    assetsFolder,
    tilesFolder,
    tileImagesFolder,
    mapsFolder,
    scalingFactor,
    tileSize,
    halfTileSize,
    charactersFolder,
    itemsFolder,
    objectsFolder,
    updateInterval,
    screenWidth,
    screenHeight,
  )
where

import System.FilePath ((</>))

fps :: Int
fps = 60

scalingFactor :: Float
scalingFactor = 4

tileSize :: Float
tileSize = 16

halfTileSize :: Float
halfTileSize = tileSize / 2

assetsFolder :: FilePath
assetsFolder = "assets"

tilesFolder :: FilePath
tilesFolder = assetsFolder </> "tiles"

imagesFolder :: FilePath
imagesFolder = assetsFolder </> "images"

tileImagesFolder :: FilePath
tileImagesFolder = imagesFolder </> "tiles"

mapsFolder :: FilePath
mapsFolder = assetsFolder </> "maps"

charactersFolder :: FilePath
charactersFolder = imagesFolder </> "characters"

itemsFolder :: FilePath
itemsFolder = imagesFolder </> "items"

objectsFolder :: FilePath
objectsFolder = imagesFolder </> "objects"

updateInterval :: Float
updateInterval = 0.5

screenWidth, screenHeight :: Int
(screenWidth, screenHeight) = (1200, 1000)
